pulp_container/app/registry_api.py

```diff
@@ -1009,3 +1049,61 @@ def get_response_data(signatures):
             }
             data.append(signature)
         return {"signatures": data}
+
+    def put(self, request, path, pk):
+        """Create a new signature from the received data."""
+        _, repository = self.get_dr_push(request, path)
+
+        try:
+            manifest = models.Manifest.objects.get(digest=pk)
+        except models.Manifest.DoesNotExist:
+            raise ManifestNotFound(reference=pk)
+
+        signature_payload = request.META["wsgi.input"].read(SIGNATURE_PAYLOAD_MAX_SIZE)
+        signature_raw = base64.b64decode(json.loads(signature_payload)["content"])
+        signature_json = extract_data_from_signature(signature_raw, manifest.digest)
+        if signature_json is None:
+            raise ManifestSignatureInvalid(digest=pk)
+
+        sig_digest = hashlib.sha256(signature_raw).hexdigest()
+        signature = models.ManifestSignature(
+            name=f"{manifest.digest}@{sig_digest[:32]}",
+            digest=f"sha256:{sig_digest}",
+            type=SIGNATURE_TYPE.ATOMIC_SHORT,
+            key_id=signature_json["signing_key_id"],
+            timestamp=signature_json["signature_timestamp"],
+            creator=signature_json["optional"].get("creator"),
+            data=base64.b64encode(signature_raw).decode(),
+            signed_manifest=manifest,
+        )
+        try:
+            signature.save()
+        except IntegrityError:
```

Comment:
reading this code it feels like integrity error should not happen because existing signatures are not getting (re)uploaded, but let's play safe https://github.com/containers/image/blob/main/docker/docker_image_dest.go#L654

Reply:
I will keep the try-except block.